corehq/form_processor/document_stores.py

```diff
@@ -60,10 +60,17 @@ def __init__(self, domain, case_type=None):
         self.domain = domain
         self.case_type = case_type
 
-    def get_document(self, doc_id):
+    def get_document(self, doc_id, *, external_id=None):
         try:
-            return CommCareCase.objects.get_case(doc_id, self.domain).to_json()
-        except CaseNotFound as e:
+            if external_id is None:
+                return CommCareCase.objects.get_case(doc_id, self.domain).to_json()
+            else:
+                return CommCareCase.objects.get_case_by_external_id(
+                    self.domain,
+                    external_id,
+                    raise_multiple=True
+                ).to_json()
+        except (CaseNotFound, CommCareCase.MultipleObjectsReturned) as e:
```

Comment:
Seems off to say that a document was not found in case of multiple matches.
I agree that we could raise an error in case of multiple matches, though it should be communicated as such.

Reply:
In the context of UCRs this makes sense. In the case of the generic API this will be surfaced to the user in the logs and the it should indicate the difference between the two errors.

Reply:
nit: I personally prefer to be really explicit about the area of code raising exception along with the expected exception.

```
if external_id is None:
    # this could be a new function _get_case_by_case_id
    try:
        case = CommCareCase.objects.get_case(doc_id, self.domain).to_json()
    except CaseNotFound as e:
        raise DocumentNotFoundError(e)
else:
    # this could be a new function _get_case_by_external_id
    try:
        case = CommCareCase.objects.get_case_by_external_id(
            self.domain,
            external_id,
            raise_multiple=True
        )
    except CommCareCase.MultipleObjectsReturned as e:
        # this is new and would need to be added. Thought of a new exception should be used here
        # defined in HQ instead of the one provided by Django
        raise MultipleDocumentsFoundError(e)
    else:            
        if case:
            return case.to_json()
        raise DocumentNotFoundError(f"external_id: '{external_id}'")
```